src/ufuncs.c

```diff
@@ -1595,6 +1607,63 @@ static void relate_pattern_func(char** args, npy_intp* dimensions, npy_intp* ste
 }
 static PyUFuncGenericFunction relate_pattern_funcs[1] = {&relate_pattern_func};
 
+#if GEOS_SINCE_3_6_0
+static char set_precision_dtypes[4] = {NPY_OBJECT, NPY_DOUBLE, NPY_BOOL, NPY_OBJECT};
+static void set_precision_func(char** args, npy_intp* dimensions, npy_intp* steps,
+                               void* data) {
+  GEOSGeometry* in1 = NULL;
+  GEOSGeometry** geom_arr;
+
+  CHECK_NO_INPLACE_OUTPUT(4);
+
+  // allocate a temporary array to store output GEOSGeometry objects
+  geom_arr = malloc(sizeof(void*) * dimensions[0]);
+  CHECK_ALLOC(geom_arr);
+
+  GEOS_INIT_THREADS;
+
+  QUATERNARY_LOOP {
+    // get the geometry: return on error
+    if (!get_geom(*(GeometryObject**)ip1, &in1)) {
+      errstate = PGERR_NOT_A_GEOMETRY;
+      destroy_geom_arr(ctx, geom_arr, i - 1);
+      break;
+    }
+    // grid size
+    double in2 = *(double*)ip2;
+    // preserve topology
+    npy_bool in3 = *(npy_bool*)ip3;
+    // flags:
+    // GEOS_PREC_NO_TOPO (1<<0): if set, do not try to preserve topology
+    // GEOS_PREC_KEEP_COLLAPSED  (1<<1): Not used because uncollapsed geometries are
+    // invalid and will not be retained in GEOS >= 3.9 anyway.
+    int flags = (!in3) << 0;
```

Comment:
As the geos c api defines these as macros, it would be more readable to do something like:

```
int flags = 0 ? in3 : GEOS_PREC_NO_TOPO;
```

Reply:
I think you meant `int flags = in3 ? 0 : GEOS_PREC_NO_TOPO;`